src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/removepolicy/TransportRemovePolicyAction.kt

```diff
@@ -86,63 +94,100 @@ class TransportRemovePolicyAction @Inject constructor(
+            indicesToRemove.forEach { multiGetReq.add(INDEX_MANAGEMENT_INDEX, it.key) }
+
+            client.multiGet(multiGetReq, object : ActionListener<MultiGetResponse> {
+                override fun onResponse(response: MultiGetResponse) {
+                    // config index may not be initialized
+                    val f = response.responses.first()
+                    if (f.isFailed && f.failure.failure is IndexNotFoundException) {
+                        indicesToRemove.forEach { (uuid, name) ->
+                            failedIndices.add(FailedIndex(name, uuid, "This index does not have a policy to remove"))
+                        }
+                        actionListener.onResponse(ISMStatusResponse(0, failedIndices))
+                        return
+                    }
+
+                    response.forEach {
+                        if (!it.response.isExists) {
+                            val docId = it.id // docId is managed index uuid
+                            failedIndices.add(FailedIndex(indicesToRemove[docId] as String, docId,
+                                    "This index does not have a policy to remove"))
+                            indicesToRemove.remove(docId)
+                        }
+                    }
+
+                    removeManagedIndex()
+                }
+
+                override fun onFailure(t: Exception) {
+                    actionListener.onFailure(t)
+                }
+            })
+        }
+
         @Suppress("SpreadOperator") // There is no way around dealing with java vararg without spread operator.
-        fun processResponse(clusterStateResponse: ClusterStateResponse) {
-            val state = clusterStateResponse.state
-            populateLists(state)
-
-            if (indicesToRemovePolicyFrom.isNotEmpty()) {
-                val updateSettingsRequest = UpdateSettingsRequest()
-                    .indices(*indicesToRemovePolicyFrom.map { it.name }.toTypedArray())
-                    .settings(Settings.builder().putNull(ManagedIndexSettings.POLICY_ID.key))
-
-                try {
-                    client.execute(UpdateSettingsAction.INSTANCE, updateSettingsRequest,
-                        object : ActionListener<AcknowledgedResponse> {
-                            override fun onResponse(response: AcknowledgedResponse) {
-                                if (response.isAcknowledged) {
-                                    updated = indicesToRemovePolicyFrom.size
-                                } else {
-                                    updated = 0
-                                    failedIndices.addAll(indicesToRemovePolicyFrom.map {
-                                        FailedIndex(it.name, it.uuid, "Failed to remove policy")
-                                    })
-                                }
-
-                                actionListener.onResponse(ISMStatusResponse(updated, failedIndices))
+        fun removeManagedIndex() {
```

Comment:
NP: Renaming this to be `removeManagedIndices` would better convey that the removal can be performed on more than one index at a time. The same goes for `createManagedIndex` in `TransportAddPolicyAction`.

Reply:
renamed